nixos/modules/config/users-groups.nix

```
@@ -893,29 +893,46 @@ in
         else
           ""; # keep around for backwards compatibility
 
-      systemd.services.linger-users = lib.mkIf ((length lingeringUsers) > 0) {
+      systemd.services.linger-users = {
         wantedBy = [ "multi-user.target" ];
         after = [ "systemd-logind.service" ];
         requires = [ "systemd-logind.service" ];
 
         script =
           let
             lingerDir = "/var/lib/systemd/linger";
-            lingeringUsersFile = builtins.toFile "lingering-users" (
-              concatStrings (map (s: "${s}\n") (sort (a: b: a < b) lingeringUsers))
-            ); # this sorting is important for `comm` to work correctly
+            userPartition = lib.lists.partition (u: u.linger) (builtins.attrValues cfg.users);
+            lingeringUserNames = map (u: u.name) userPartition.right;
+            nonLingeringUserNames = map (u: u.name) userPartition.wrong;
           in
           ''
+            ${lib.strings.toShellVars { inherit lingeringUserNames nonLingeringUserNames; }}
+
+            user_configured () {
+                # Use `id` to check if the user exists rather than checking the
+                # NixOS configuration, as it may be that the user has been
+                # manually configured, which is permitted if users.mutableUsers
+                # is false (the default).
```

```suggestion
                # is true (the default).
```

This one is just inverted.